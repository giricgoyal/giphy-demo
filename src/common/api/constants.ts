export const GIPHY_URL_BASE = 'https://api.giphy.com/v1'
export const GIPHY_API_KEY = 'tVaJe9QRTL6VZp9xhBkogbNWFTI9hYnJ'
export const GIPHY_API_OFFSET_VALUE = 0
export const GIPHY_API_LIMIT_VALUE = 25
