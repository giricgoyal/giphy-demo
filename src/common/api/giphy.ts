import axios from 'axios'
import { GIF_DATA, API_PARAMS } from '../../types'
import { GIPHY_API_KEY, GIPHY_URL_BASE } from './constants'

type AXIOS_RESPONSE = {
    data: GIF_DATA
}

export const getUrl = (path: string, resource = 'gifs'): string => `${GIPHY_URL_BASE}/${resource}/${path}`

const getData = (url: string, params: API_PARAMS): Promise<AXIOS_RESPONSE> => {
    return axios.get(url, {
        params: {
            api_key: GIPHY_API_KEY,
            ...params,
        },
    })
}

export const search = async ({ q, offset, limit }: API_PARAMS): Promise<GIF_DATA> => {
    try {
        const url = getUrl('search')
        const result: AXIOS_RESPONSE = await getData(url, {
            q,
            offset,
            limit,
        })

        return result.data
    } catch (error) {
        throw error
    }
}

export const trending = async ({ offset, limit }: API_PARAMS): Promise<GIF_DATA> => {
    try {
        const url = getUrl('trending')
        const result = await getData(url, {
            offset,
            limit,
        })

        return result.data
    } catch (error) {
        throw error
    }
}
