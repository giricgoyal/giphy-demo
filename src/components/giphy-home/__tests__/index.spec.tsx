import React from 'react'
import { act, fireEvent, render, waitFor, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import GiphyHome from '../index'
import * as apis from '../../../common/api/giphy'
import { setupIntersectionObserverMock } from '../../../mocks'

describe('<GiphyHome />', () => {
    let container, getByTestId
    let trendingSpy, searchSpy

    beforeEach(() => {
        setupIntersectionObserverMock()
        trendingSpy = jest.spyOn(apis, 'trending').mockImplementation(() =>
            Promise.resolve({
                data: [
                    {
                        id: '1',
                        images: {
                            original: {
                                url: 'https://www.superman.com/img.gif',
                            },
                            '480w_still': {
                                url: 'https://www.superman.com/img.gif',
                            },
                        },
                        title: 'test1',
                    },
                ],
            }),
        )
        searchSpy = jest.spyOn(apis, 'search').mockImplementation(() =>
            Promise.resolve({
                data: [
                    {
                        id: '1',
                        images: {
                            original: {
                                url: 'https://www.batman.com/img.gif',
                            },
                            '480w_still': {
                                url: 'https://www.batman.com/img.gif',
                            },
                        },
                        title: 'meh',
                    },
                ],
            }),
        )
        ;({ container, getByTestId } = render(<GiphyHome />))
    })

    test('should match snapshot', async () => {
        await waitFor(() => {
            expect(container).toMatchSnapshot()
        })
    })

    test('should call the trending api', async () => {
        await waitFor(() => {
            expect(trendingSpy).toHaveBeenCalledWith({
                offset: 0,
                limit: 20,
            })
        })
    })

    test('should not render the modal', () => {
        expect(screen.queryByTestId('modal-overlay')).toEqual(null)
        expect(screen.queryByTestId('modal')).toEqual(null)
    })

    describe('When search is used', () => {
        beforeEach(() => {
            act(() => {
                fireEvent.change(getByTestId('input'), {
                    target: {
                        value: 'Meh',
                    },
                })
            })
        })

        test('should call the search api', async () => {
            await waitFor(() => {
                expect(searchSpy).toHaveBeenCalledWith({
                    q: 'Meh',
                    offset: 0,
                    limit: 20,
                })
            })
        })

        test('should match snapshot', async () => {
            await waitFor(() => {
                expect(screen.getByTestId('giphy-home')).toMatchSnapshot()
            })
        })
    })

    describe('Modal', () => {
        describe('When an item is clicked', () => {
            beforeEach(() => {
                fireEvent.click(screen.getAllByTestId('grid-item')[0])
            })

            test('should match snapshot', async () => {
                await waitFor(() => {
                    expect(screen.getByTestId('modal')).toMatchSnapshot()
                })
            })

            describe('When modal is closed', () => {
                beforeEach(() => {
                    fireEvent.click(screen.getByTestId('modal-close'))
                })

                test('should not render the modal', () => {
                    expect(screen.queryByTestId('modal-overlay')).toEqual(null)
                    expect(screen.queryByTestId('modal')).toEqual(null)
                })
            })
        })
    })

    // describe.skip('When tending api resolves', () => {
    //     beforeEach(() => {
    //         trendingSpy.mockImplementation(() =>
    //             Promise.resolve({
    //                 data: [],
    //                 pagination: {},
    //             }),
    //         )
    //     })

    //     test('should match snapshot', () => {})
    // })
})
