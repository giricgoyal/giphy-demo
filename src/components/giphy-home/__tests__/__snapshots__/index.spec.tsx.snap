// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiphyHome /> Modal When an item is clicked should match snapshot 1`] = `
<div
  class="modal"
  data-testid="modal"
>
  <div
    class="modal__header"
  >
    <div
      class="modal__header__title"
    >
      test1
    </div>
    <svg
      class="modal__header__close icon"
      data-testid="modal-close"
    >
      <use
        xlink:href="#undefined"
      />
    </svg>
  </div>
  <div
    class="modal__body"
  >
    <div
      class="renditions"
    >
      <div
        class="renditions__sidebar"
      >
        <div
          class="renditions__sidebar__item active"
          data-testid="sidebar-item"
        >
          Original
        </div>
        <div
          class="renditions__sidebar__item "
          data-testid="sidebar-item"
        >
          480 W Still
        </div>
      </div>
      <div
        class="renditions__body"
      >
        <img
          src="https://www.superman.com/img.gif"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<GiphyHome /> When search is used should match snapshot 1`] = `
<div
  class="giphy-home container"
  data-testid="giphy-home"
>
  <form
    class="search"
  >
    <input
      class="search__input form-control"
      data-testid="input"
      placeholder="Search"
      type="text"
      value="Meh"
    />
    <svg
      class="search__icon icon"
    >
      <use
        xlink:href="#undefined"
      />
    </svg>
    <svg
      class="search__clear icon"
      data-testid="clear"
    >
      <use
        xlink:href="#undefined"
      />
    </svg>
  </form>
  <div
    class="grid"
    data-testid="grid"
  >
    <div
      class="grid__header"
    >
      <span>
        Search results
      </span>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      />
      <div
        class="col"
      />
      <div
        class="col"
      />
      <div
        class="col"
      />
    </div>
    <div />
  </div>
</div>
`;

exports[`<GiphyHome /> should match snapshot 1`] = `
<div>
  <div
    class="giphy-home container"
    data-testid="giphy-home"
  >
    <form
      class="search"
    >
      <input
        class="search__input form-control"
        data-testid="input"
        placeholder="Search"
        type="text"
        value=""
      />
      <svg
        class="search__icon icon"
      >
        <use
          xlink:href="#undefined"
        />
      </svg>
      
    </form>
    <div
      class="grid"
      data-testid="grid"
    >
      <div
        class="grid__header"
      >
        <svg
          class="undefined icon"
        >
          <use
            xlink:href="#undefined"
          />
        </svg>
        <span>
          Trending
        </span>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="grid__item cursor-pointer"
            data-testid="grid-item"
          >
            <img
              class="giphy-home__image"
              src="https://www.superman.com/img.gif"
            />
          </div>
        </div>
        <div
          class="col"
        />
        <div
          class="col"
        />
        <div
          class="col"
        />
      </div>
      <div />
    </div>
  </div>
</div>
`;
